### PR TITLE
UCT/IB: Cleanup device check and add ConnectX-5 support.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: src/uct/ib/base/ib_device.c [hunk 1]
```diff
@@ -27,6 +27,16 @@ static ucs_stats_class_t uct_ib_device_stats_class = {
...
 #endif
 
+static uct_ib_device_info_t uct_ib_device_info_table[] = {
+  {0x02c9, 4099, "ConnectX-3",     UCT_IB_DEVICE_FLAG_MLX4_PRM, 10},
+  {0x02c9, 4103, "ConnectX-3 Pro", UCT_IB_DEVICE_FLAG_MLX4_PRM, 11},
+  {0x02c9, 4113, "Connect-IB",     UCT_IB_DEVICE_FLAG_MLX5_PRM, 20},
+  {0x02c9, 4115, "ConnectX-4",     UCT_IB_DEVICE_FLAG_MLX5_PRM, 30},
+  {0x02c9, 4117, "ConnectX-4 LX",  UCT_IB_DEVICE_FLAG_MLX5_PRM, 28},
+  {0x02c9, 4121, "ConnectX-5",     UCT_IB_DEVICE_FLAG_MLX5_PRM, 40},
+  {0,      0,    "Generic HCA",    0,                           0}
```
Comment (miked-mellanox) [inline src/uct/ib/base/ib_device.c]: Maybe add entry from env vars if exist.this will allow adding new card at runtime with vars w.o. Code changes.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3303/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/835/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2787/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2790/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3306/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/837/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3307/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/838/ for details.

File: src/uct/ib/base/ib_md.c [hunk 2]
```diff
@@ -990,6 +1001,83 @@ static void uct_ib_fork_warn_enable()
...
+
+        if (nfields >= 4) {
+            for (p = flags_str; *p != 0; ++p) {
+                if (*p == '4') {
+                    spec->flags |= UCT_IB_DEVICE_FLAG_MLX4_PRM;
+                } else if (*p == '5') {
+                    spec->flags |= UCT_IB_DEVICE_FLAG_MLX5_PRM;
+                } else {
+                    ucs_error("invalid device flag: '%c'", *p);
+                    free(flags_str);
```
Comment (brminich) [inline src/uct/ib/base/ib_md.c]: seems spec->name should also be freed

Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: yrp

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2791/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2795/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3311/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3312/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/843/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2796/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/844/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3313/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2797/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/845/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2803/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2805/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@miked-mellanox pls review, added env var to specify devices

File: src/uct/ib/base/ib_md.c [hunk 3]
```diff
@@ -114,6 +114,17 @@ static ucs_config_field_t uct_ib_md_config_table[] = {
    "Maximal memory region size to enable ODP for. 0 - disable.\n",
    ucs_offsetof(uct_ib_md_config_t, odp.max_size), UCS_CONFIG_TYPE_MEMUNITS},
 
+  {"DEVICE_SPECS", "",
+   "Array of custom device specification. Each element is a string of the following format:\n"
+   "  <vendor-id>:<part-id>[:name[:<flags>[:<priority>]]]\n"
+   "where:\n"
+   "  <vendor-id> - (mandatory) vendor id, integer or hexadecimal.\n"
+   "  <part-id>   - (mandatory) vendor part id, integer or hexadecimal.\n"
```
Comment (miked-mellanox) [inline src/uct/ib/base/ib_md.c]: vendorid/partid - is it taken from OFA (ibstat) or lspci?

Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: ofa

**Pull request merged**